File: kafka-connect-jdbc-sink/src/main/scala/com/datamountaineer/streamreactor/connect/jdbc/sink/StructFieldsDataExtractor.scala
package com.datamountaineer.streamreactor.connect.jdbc.sink

import org.apache.kafka.connect.data.{Field, Schema, Struct}

import scala.collection.JavaConversions._

case class StructFieldsDataExtractor(includeAllFields: Boolean, fieldsAliasMap: Map[String, String]) {

  def get(struct: Struct): Seq[(String, PreparedStatementBinder)] = {
    val schema = struct.schema()
    val fields: Seq[Field] = if (includeAllFields) schema.fields()
    else schema.fields().filter(f => fieldsAliasMap.contains(f.name()))

    val fieldsAndValues = fields.flatMap { case field =>
      getFieldValue(field, struct).map(value => fieldsAliasMap.getOrElse(field.name(), field.name()) -> value)
    }
    fieldsAndValues
  }

  private def getFieldValue(field: Field, struct: Struct): Option[PreparedStatementBinder] = {
    Option(struct.get(field)) match {
      case None => None
      case Some(value) =>
        val fieldName = field.name()
        val value = field.schema() match {
          case Schema.BOOLEAN_SCHEMA | Schema.OPTIONAL_BOOLEAN_SCHEMA => BooleanPreparedStatementBinder(struct.getBoolean(fieldName))
          case Schema.BYTES_SCHEMA | Schema.OPTIONAL_BYTES_SCHEMA => BytesPreparedStatementBinder(struct.getBytes(fieldName))
          case Schema.FLOAT32_SCHEMA | Schema.OPTIONAL_FLOAT32_SCHEMA => FloatPreparedStatementBinder(struct.getFloat32(fieldName))
          case Schema.FLOAT64_SCHEMA | Schema.OPTIONAL_FLOAT64_SCHEMA => DoublePreparedStatementBinder(struct.getFloat64(fieldName))
          case Schema.INT8_SCHEMA | Schema.OPTIONAL_INT8_SCHEMA => BytePreparedStatementBinder(struct.getInt8(fieldName))
          case Schema.INT16_SCHEMA | Schema.OPTIONAL_INT16_SCHEMA => ShortPreparedStatementBinder(struct.getInt16(fieldName))
          case Schema.INT32_SCHEMA | Schema.OPTIONAL_INT32_SCHEMA => IntPreparedStatementBinder(struct.getInt32(fieldName))
          case Schema.INT64_SCHEMA | Schema.OPTIONAL_INT64_SCHEMA => LongPreparedStatementBinder(struct.getInt64(fieldName))
          case Schema.STRING_SCHEMA | Schema.OPTIONAL_STRING_SCHEMA => StringPreparedStatementBinder(struct.getString(fieldName))
        }
        Some(value)
    }
  }
}

File: kafka-connect-jdbc-sink/src/main/java/com/datamountaineer/streamreactor/connect/jdbc/sink/StructFieldsDataExtractor.java
/**
 * Copyright 2015 Datamountaineer.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


package com.datamountaineer.streamreactor.connect.jdbc.sink;

import com.datamountaineer.streamreactor.connect.Pair;
import com.datamountaineer.streamreactor.connect.jdbc.sink.binders.*;
import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Lists;
import org.apache.kafka.connect.data.Field;
import org.apache.kafka.connect.data.Schema;
import org.apache.kafka.connect.data.Struct;

import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * This class holds the a mappings of fields to extract from
 * a Connect Struct record.
 *
 * Used to building mappings fro Struct records to JDBC binding statements.
 *
 * For example, if the struct contains a string field which is part of the aliasMap .i.e.
 * set in configuration to be write to the target, it will return a StringPreparedStatementBinder
 * (statement.setString(index, value))
 * */
public class StructFieldsDataExtractor {
    private final boolean includeAllFields;
    private final Map<String, String> fieldsAliasMap;

    public StructFieldsDataExtractor(boolean includeAllFields, Map<String, String> fieldsAliasMap) {
        this.includeAllFields = includeAllFields;
        this.fieldsAliasMap = fieldsAliasMap;
    }

    /**
     * For a Struct record build list of PreparedStatementBinders based on the structs fields.
     *
     * @param struct A Connect Struct record to extract fields from
     * @return A list if mappings of field name to PreparedStatementBinder for each field in the struct.
     * */
    public List<Pair<String, PreparedStatementBinder>> get(final Struct struct) {
        //get the records schema
        final Schema schema = struct.schema();
        final Collection<Field> fields;

        //if all fields take all the fields from the schema else filter out the fields based on the provided map
        if (includeAllFields)
            fields = schema.fields();
        else {
            fields = Collections2.filter(schema.fields(), new Predicate<Field>() {
                @Override
                public boolean apply(Field input) {
                    //filter for field name in map
                    return fieldsAliasMap.containsKey(input.name()) || includeAllFields;
                }
                @Override
                public boolean equals(Object object) {
                    return false;
                }
            });
        }

        //For each field convert the Connect schema type and extract the value, add to list and return
        final List<Pair<String, PreparedStatementBinder>> binders = Lists.newArrayList();
        for (final Field field : fields) {
            //get the correct binder based on the struct fields type
            final PreparedStatementBinder binder = getFieldValue(field, struct);
            if (binder != null) {
                String fieldName;
                if (fieldsAliasMap.containsKey(field.name())) {
                    fieldName = fieldsAliasMap.get(field.name());
                }
                else {
                    fieldName = field.name();
                }
                binders.add(new Pair(fieldName, binder));
            }
        }
        return binders;
    }

    /**
     * Return a PreparedStatementBinder for a struct fields.
     *
     * @param field The struct field to get the binder for.
     * @param struct The struct which the field belongs to.
     * @return A PreparedStatementBinder for the field.
     * */
    private PreparedStatementBinder getFieldValue(final Field field, final Struct struct) {
        final Object value = struct.get(field);
        if (value == null) {
            return null;
        }

        final String fieldName = field.name();

        //match on fields schema type to find the correct casting.
        PreparedStatementBinder binder;
        switch (field.schema().type()) {
            case INT8:
                binder = new BytePreparedStatementBinder(struct.getInt8(fieldName));
                break;
            case INT16:
                binder = new ShortPreparedStatementBinder(struct.getInt16(fieldName));
                break;
            case INT32:
                binder = new IntPreparedStatementBinder(struct.getInt32(fieldName));
                break;
            case INT64:
                binder = new LongPreparedStatementBinder(struct.getInt64(fieldName));
                break;
            case FLOAT32:
                binder = new FloatPreparedStatementBinder(struct.getFloat32(fieldName));
                break;
            case FLOAT64:
                binder = new DoublePreparedStatementBinder(struct.getFloat64(fieldName));
                break;
            case BOOLEAN:
                binder = new BooleanPreparedStatementBinder(struct.getBoolean(fieldName));
                break;
            case STRING:
                binder = new StringPreparedStatementBinder(struct.getString(fieldName));
                break;
            case BYTES:
                binder = new BytesPreparedStatementBinder(struct.getBytes(fieldName));
                break;
            default:
                throw new IllegalArgumentException("Following schema type " + struct.schema().type() + " is not supported");
        }
        return binder;
    }
}

